From 4c35149a8696a0b49cfbdc71a6333f923fb1b875 Mon Sep 17 00:00:00 2001
From: ohfp <1813007-ohfp@users.noreply.gitlab.com>
Date: Sat, 20 Feb 2021 13:35:56 +0100
Subject: [PATCH 2/2] update neon so it does not fail due to type issues on
 arch

---
 bin/build-electron | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/bin/build-electron b/bin/build-electron
index 47e19be..e8b2855 100755
--- a/bin/build-electron
+++ b/bin/build-electron
@@ -105,10 +105,10 @@ fi
 
         if [ "${BUILD_TYPE}" = "debug" ]
         then
-            /usr/bin/gn gen -C "${OUTPUT_DIR}"/Debug "--args=custom_toolchain=\"//build/toolchain/linux/unbundle:default\" host_toolchain=\"//build/toolchain/linux/unbundle:default\" is_clang=true use_custom_libcxx=false rtc_build_examples=false rtc_build_tools=false rtc_include_tests=false rtc_enable_protobuf=false rtc_use_x11=false clang_use_chrome_plugins=false"
+            /usr/bin/gn gen -C "${OUTPUT_DIR}"/Debug "--args=custom_toolchain=\"//build/toolchain/linux/unbundle:default\" host_toolchain=\"//build/toolchain/linux/unbundle:default\" target_cpu=\"arm64\" is_clang=true use_custom_libcxx=false rtc_build_examples=false rtc_build_tools=false rtc_include_tests=false rtc_enable_protobuf=false rtc_use_x11=false clang_use_chrome_plugins=false"
             /usr/bin/ninja -C "${OUTPUT_DIR}"/Debug
         else
-            /usr/bin/gn gen -C "${OUTPUT_DIR}"/Release "--args=custom_toolchain=\"//build/toolchain/linux/unbundle:default\" host_toolchain=\"//build/toolchain/linux/unbundle:default\" is_clang=true use_custom_libcxx=false rtc_build_examples=false rtc_build_tools=false rtc_include_tests=false rtc_enable_protobuf=false rtc_use_x11=false is_debug=false clang_use_chrome_plugins=false"
+            /usr/bin/gn gen -C "${OUTPUT_DIR}"/Release "--args=custom_toolchain=\"//build/toolchain/linux/unbundle:default\" host_toolchain=\"//build/toolchain/linux/unbundle:default\" target_cpu=\"arm64\" is_clang=true use_custom_libcxx=false rtc_build_examples=false rtc_build_tools=false rtc_include_tests=false rtc_enable_protobuf=false rtc_use_x11=false is_debug=false clang_use_chrome_plugins=false"
             /usr/bin/ninja -C "${OUTPUT_DIR}"/Release
         fi
     else
@@ -134,11 +134,14 @@ fi
     then
         config_arch=x64
     fi
+
+    cargo update -p neon
+
     if [ "${BUILD_TYPE}" = "debug" ]
     then
-        npm_config_arch=${config_arch} npm_config_target_arch=${config_arch} npm_config_disturl=https://atom.io/download/electron npm_config_runtime=electron npm_config_target=8.2.5 npm_config_build_from_source=true npm_config_devdir=~/.electron-gyp OUTPUT_DIR="${OUTPUT_DIR}" cargo build --features electron
+        npm_config_arch=${config_arch} npm_config_target_arch=${config_arch} npm_config_disturl=https://atom.io/download/electron npm_config_runtime=electron npm_config_target=8.2.5 npm_config_build_from_source=true npm_config_devdir=~/.electron-gyp cargo build --target=aarch64-unknown-linux-gnu --features electron
     else
-        npm_config_arch=${config_arch} npm_config_target_arch=${config_arch} npm_config_disturl=https://atom.io/download/electron npm_config_runtime=electron npm_config_target=8.2.5 npm_config_build_from_source=true npm_config_devdir=~/.electron-gyp RUSTFLAGS="-C link-arg=-s" OUTPUT_DIR="${OUTPUT_DIR}" cargo build --features electron --release
+        npm_config_arch=${config_arch} npm_config_target_arch=${config_arch} npm_config_disturl=https://atom.io/download/electron npm_config_runtime=electron npm_config_target=8.2.5 npm_config_build_from_source=true npm_config_devdir=~/.electron-gyp RUSTFLAGS="-C link-arg=-s" OUTPUT_DIR="${OUTPUT_DIR}" cargo build --target=aarch64-unknown-linux-gnu --features electron --release
     fi
 
     if [ $DEFAULT_PLATFORM = "darwin" ]
@@ -152,6 +155,6 @@ fi
     elif [ $DEFAULT_PLATFORM = "linux" ] || [ $DEFAULT_PLATFORM = "linux-aarch64" ]
     then
         mkdir -p ../node/build/linux
-        cp -f target/${BUILD_TYPE}/libringrtc.so ../node/build/linux/libringrtc.node
+        cp -f target/aarch64-unknown-linux-gnu/${BUILD_TYPE}/libringrtc.so ../node/build/linux/libringrtc.node
     fi
 )
-- 
2.30.2

