diff --git a/BUILD.gn b/BUILD.gn
index e2d3b06cef..b77086bd1d 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -414,10 +414,12 @@ config("common_config") {
 }
 
 config("common_objc") {
-  libs = [ "Foundation.framework" ]
+  if (is_ios || is_mac) {
+    libs = [ "Foundation.framework" ]
 
-  if (rtc_use_metal_rendering) {
-    defines = [ "RTC_SUPPORTS_METAL" ]
+    if (rtc_use_metal_rendering) {
+      defines = [ "RTC_SUPPORTS_METAL" ]
+    }
   }
 }
 
